build-logic/src/main/groovy/net/fabricmc/yarn/buildlogic/task/JavadocLintTask.groovy

```diff
@@ -0,0 +1,103 @@
+
+import cuchaz.enigma.ProgressListener
+import cuchaz.enigma.translation.mapping.EntryMapping
+import cuchaz.enigma.translation.mapping.serde.MappingFileNameFormat
+import cuchaz.enigma.translation.mapping.serde.MappingSaveParameters
+import cuchaz.enigma.translation.mapping.serde.enigma.EnigmaMappingsReader
+import cuchaz.enigma.translation.mapping.tree.EntryTree
+import cuchaz.enigma.translation.representation.entry.Entry
+import cuchaz.enigma.translation.representation.entry.LocalVariableEntry
+import cuchaz.enigma.translation.representation.entry.MethodEntry
+import groovy.transform.CompileStatic
+import org.gradle.api.DefaultTask
+import org.gradle.api.GradleException
+import org.gradle.api.file.DirectoryProperty
+import org.gradle.api.tasks.InputDirectory
+import org.gradle.api.tasks.TaskAction
+
+@CompileStatic
+class JavadocLintTask extends DefaultTask {
+    private static final String PARAMETER_TAG_START = "@param "
+
+    @InputDirectory
+    DirectoryProperty mappingDirectory = project.objects.directoryProperty()
+
+    JavadocLintTask() {
+        // Only run when inputs have changed
+        outputs.upToDateWhen { true }
+    }
+
+    @TaskAction
+    def run() {
+        def saveParameters = new MappingSaveParameters(MappingFileNameFormat.BY_DEOBF)
+        def directory = mappingDirectory.getAsFile().get().toPath()
+        EntryTree<EntryMapping> mappings = EnigmaMappingsReader.DIRECTORY.read(directory, ProgressListener.none(), saveParameters)
+        def errors = []
+
+        mappings.allEntries.parallel().forEach { Entry<?> entry ->
+            EntryMapping mapping = mappings.get(entry)
+            String javadoc = mapping.javadoc
+
+            if (javadoc != null && !javadoc.isEmpty()) {
+                def localErrors = []
+
+                if (entry instanceof LocalVariableEntry && entry.argument) {
+                    if (javadoc.endsWith(".")) {
+                        localErrors.add("parameter javadoc ends with '.'")
+                    }
+
+                    if (javadoc.charAt(0).isUpperCase()) {
+                        String word = getFirstWord(javadoc)
+
+                        // ignore single-letter "words" (like X or Z)
+                        if (word.length() > 1) {
+                            localErrors.add("parameter javadoc starts with uppercase word '${word}'")
+                        }
+                    }
+                }
+
+                if (entry instanceof MethodEntry) {
+                    // check for non-type parameters declared as @param
+                    if (javadoc.contains(PARAMETER_TAG_START)) {
+                        int paramIndex = javadoc.indexOf(PARAMETER_TAG_START) + PARAMETER_TAG_START.length()
+
+                        if (javadoc.length() <= paramIndex || javadoc.charAt(paramIndex) != '<' as char) {
+                            localErrors.add("method contains parameter docs, which should be on the parameter itself")
```

Comment:
Rip my bad, you have considered this case, good

Reply:
Hmm, there should be no need for `== paramIndex` case, maybe just use `javadoc.length() < paramIndex`

Reply:
Wouldn't that technically be prone to throwing an index out of bounds exception when the doc ends with `@param `? (Very unlikely, but still possible)